created remaining backend files

Arushi55 opened the pull request:
Created all necessary backend files and folders.
The database connection part may be wrong, will need to review that once.
Have four controller and accessor methods (one for each HTTP request) that will now need to be made accessible from the front end.

Arushi55 committed:
created remaining backend files

created remaining backend folders and files
connection to db is questionable at best, needs to be reviewed
installed mongoose

ethanszeto commented on package.json:
```diff
@@ -9,6 +9,6 @@
     "start-database": "cd database && npm start"
   },
-  "author": "",
+  "author": "Arushi Aggarwal",
```
💯

RaisaBh commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+import Accessor from "../db_accessors/accessor.js";
+
+export default class Controller {
+    static async getBlogs(req, res) {
```
Good use of naming

RaisaBh commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+            return messages;
+        } catch (e) {
+            console.log("Something bad happened due to:", e);
+        }
```
I know this was something I originally wrote but I think it would be good practice if all the console.log messages were consistent so change line 10 to match the other console.log messages.

RaisaBh commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+import Accessor from "../db_accessors/accessor.js";
```
Overall this was very good! Your code had good naming conventions and the code is well written.

Something that would be very helpful is comments. You should add comments saying what this file is for and what each function does so that students who are reading this for the first time know the purpose of each function. (I know I didn't do this but I should have and we are here to improve)

RaisaBh commented on backend/app/db/connection.js:
```diff
@@ -0,0 +1,23 @@
+
+            return mongoose.connection;
+        } catch (e) {
+            console.log("Database failure")
```
Nitpicky but missing a semicolon here.

RaisaBh commented on backend/app/db/connection.js:
```diff
@@ -0,0 +1,23 @@
+import { config as dotenvConfig } from 'dotenv';
+import mongoose from 'mongoose';
+
```
Very good just add a comment explaining what this file does.

RaisaBh commented on database/package.json:
```diff
@@ -5,7 +5,7 @@
     "start": "docker compose up -d"
   },
-  "author": "",
+  "author": "Arushi Aggarwal",
```
Lol I didn't even realize I forgot to put an author but thanks for fixing that

RaisaBh commented on backend/app/routes/defaultRoutes.js:
```diff
@@ -1,10 +1,15 @@
 import express from "express";
```
Once again good but add comments.

RaisaBh commented on backend/app/models/blog.js:
```diff
@@ -0,0 +1,19 @@
+import mongoose from "mongoose";
```
Good good but add comments explaining what a Schema is and the difference between a database and a collection

RaisaBh commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+            return messages;
+        } catch (e) {
+            console.log("Something bad will happen");
+        }
```
I know that once again I probably wrote these console.logs originally so sorry about that but it would be good practice to have all the console.log messages be the same and have them all show what the error message is

RaisaBh commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+import Blog from "../models/blog.js";
+
+export default class Accessor {
+    static async getAllBlogs() {
```
Good names for your functions!

RaisaBh commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+import Connection from "../db/connection.js";
```
Great job with this section I think again just add comments explaining everything!

RaisaBh reviewed (commented):
Overall great job! This code is well written and very clean! Just to reiterate the biggest feedback I gave across your code is that comments are good practice to explain how things work. Especially in this situation where first time students are given a bunch of code it would be helpful to have comments explaining things so that if they look at the code on their own they know what's happening

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+export default class Controller {
+    static async getBlogs(req, res) {
+        try {
+            const messages = await Accessor.getAllBlogs();
```
Rename const, no longer a message board. Same for all files

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+    static async postBlog(req, res) {
+        try {
+            await Accessor.postBlog(req.body);
+            res.redirect("/");
```
You can't redirect origin from an API. You can think of this code living under localhost:8001, a redirect will take you to localhost:8001/, not localhost:3000/...

You can do this by supplying a "redirectURL" as a field in the JSON response so that the frontend service can direct itself there after the request had been processed. However this is not good practice since this introduces tight coupling of backend and frontend

Arushi55 replied:
Do I need a redirect here at all? Can I just return and have the front end handle the redirect?

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+            res.json(messages);
+            return messages;
+        } catch (e) {
+            console.log("Something bad happened due to:", e);
```
Consistent error naming/handling would be best. Whether it's a console.log or sending a 400/500 status code res, standardize it

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,43 @@
+        try {
+            const id = req.params._id;
+            const deletedBlog = await Accessor.deleteBlogById(id);
+            res.redirect("/");
```
Again, can't redirect origin from an API

ethanszeto commented on backend/app/db/connection.js:
```diff
@@ -0,0 +1,23 @@
+            const { MONGO_USERNAME, MONGO_PASSWORD, MONGO_HOSTNAME, MONGODB_PORT } = process.env;
+            const DATABASE_URL = `mongodb://${MONGO_USERNAME}:${MONGO_PASSWORD}@${MONGO_HOSTNAME}:${MONGODB_PORT}`;
+
+            mongoose.connect(DATABASE_URL, {
```
You only need to destructure the mongodb connection string from the .env file. Then assign it to DATABSE_URL or just use it directly on the connect call.

ethanszeto commented on backend/app/db/connection.js:
```diff
@@ -0,0 +1,23 @@
+            mongoose.connect(DATABASE_URL, {
+                maxPoolSize: 50,
+                socketTimeoutMS: 2500,
+                dbName: db
```
I believe this is not a necessary field -- you can just connect to cluster, not to a specific db in general and I believe it's fine

ethanszeto commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+export default class Accessor {
+    static async getAllBlogs() {
+        try {
+            await Connection.open("teaching_proj");
```
Param not necessary if connection is to cluster, not to a db

ethanszeto commented on backend/app/models/blog.js:
```diff
@@ -0,0 +1,19 @@
+
+const BlogSchema = new Schema(
+    {
+        title: { type: String, required: true },
```
Unique constraint anywhere?

ethanszeto commented on backend/app/models/blog.js:
```diff
@@ -0,0 +1,19 @@
+        content: { type: String, required: true }
+    },
+    {
+        collection: 'proj_collection'
```
Please choose a more relevant collection name

ethanszeto commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+    static async updateBlogById(id, updatedInfo) {
+        try {
+            await Connection.open("teaching-proj");
+            const msge = await Blog.findOneAndUpdate({ _id: id }, { $set: updatedInfo }, { new: true, runValidators: true });
```
If you're not finding by index, you can use findByIdAndUpdate, for this specific use case.

https://mongoosejs.com/docs/api/model.html#Model.findByIdAndUpdate()

If you end up searching by, Title for example, and all Titles are unique, then you should use findOneAndUpdate

ethanszeto commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,44 @@
+    static async deleteBlogById(id) {
+        try {
+            await Connection.open("teaching_proj");
+            const msge = await Message.findOneAndDelete({ _id: id });
```
https://mongoosejs.com/docs/api/model.html#Model.findByIdAndDelete()

Arushi55 committed:
pull request changes

Added comments explaining all files
Changed connection.js to be cleaner
Removed redirection from API

ethanszeto commented on backend/app/app.js:
```diff
@@ -1,3 +1,7 @@
+/* This file is what sets up some of the necessary packages and 
+then directs the backend to the routes file. Routes is where 
+information can be accessed from. */
```
app is actually the definition of the web server, and related services that you want to mount to the server

even routes can be considered an additional 'service' since under the hood, that is how Express works. (See `app.use` )

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,55 @@
+/* This is the controller file. Its functions are called when 
+a route is navigated to. These functions in turn call upon the 
+database accessor to get the requested information out of the 
+database to return to the front end. */
```
Bridge of comunication between Route and DB accessor.

Route <--> Controller <-->  DB Accessor

ethanszeto commented on backend/app/controllers/controller.js:
```diff
@@ -0,0 +1,55 @@
+            res.json(blogs);
+            return blogs;
+        } catch (e) {
+            console.log("Failed due to:", e);
```
For the sake of the server not crashing, every branch of code should result in some sort of HTTP response -- plus, when thinking about frontend, you can set up listeners for a 200 or !200 response, if you set status code to error out here as a res.json

ethanszeto replied:
When an HTTP request is not met with a direct response, the server hangs and times out / crashes

ethanszeto commented on backend/app/db/connection.js:
sweet

ethanszeto commented on backend/app/db_accessors/accessor.js:
```diff
@@ -0,0 +1,53 @@
+/* This is the accessor file. It connects to the database and 
+retrieves the requested information from the database. This is 
+done using MongoDB functions. */
```
MongoDB, specifically Mongoose functions, but I think it's fine to leave it here since we don't need to explain what the Mongoose ODM is

ethanszeto commented on backend/app/routes/defaultRoutes.js:
```diff
@@ -1,10 +1,20 @@
+/* This is the routes file. When a user navigates to a specific 
+URL, this file subsequently determines which function to call to 
+set off the chain of backend function calls to get the information 
+the user is looking for. */
```
Nice

ethanszeto reviewed (changes requested):
Only one legitimate change requested, every HTTP request needs to be met with a response. In every catch block, if the code ends up there, no response is sent back over the network. This means that some code calls can result in a broken protocol, crashing the server. Instead of / in addition to your console.log, you can send back a JSON with error status code to signify error and still follow protocol.

This is personal opinion, but some comments are very verbose -- generally I find that people's eyes are averted by long paragraphs or sentences of text.

Other than that one change, looks good, and I will approve

Arushi55 committed:
added responses to catch statements
